File: backend/app/schemas/scenario.py
"""
CropIQ Scenario Pydantic Schemas
Defines request and response contracts for what-if simulations, sensitivity curves, and presets.
"""

from typing import Any, Dict, List, Optional
import math
from pydantic import BaseModel, Field, field_validator
from .farm import FarmInput


class ScenarioRequest(BaseModel):
    current_input: FarmInput = Field(..., description="Baseline farm observation")
    changes: Dict[str, float] = Field(
        ...,
        description="Features to modify and their new hypothetical values",
        example={"soil_moisture": 34.0},
    )
    scenario_name: Optional[str] = Field(None, example="Supplemental Irrigation")
    run_explanation: bool = Field(default=True, description="Whether to compute delta SHAP contributions")

    @field_validator("changes")
    @classmethod
    def validate_changes(cls, changes: Dict[str, float]) -> Dict[str, float]:
        if not changes and changes != {}:
            raise ValueError("Changes must be a dictionary.")

        # Reject target mutation
        if "yield" in changes:
            raise ValueError(
                "Modification of target 'yield' is strictly forbidden. "
                "Crop yield is an estimated outcome, not an input parameter."
            )

        # Reject metadata / identifier mutation
        for ident in ("field_id", "date_of_image", "raw_row_index", "source_dataset"):
            if ident in changes:
                raise ValueError(f"Modification of metadata '{ident}' is not allowed in scenarios.")

        # Reject crop_type change
        if "crop_type" in changes:
            raise ValueError(
                "Modifying crop_type inside what-if scenario is not supported. "
                "Crop substitutions represent systemic transitions."
            )

        for feat, val in changes.items():
            if val is None or math.isnan(val) or math.isinf(val):
                raise ValueError(f"Value for '{feat}' cannot be NaN or infinite.")

        return changes


class ScenarioBaseline(BaseModel):
    input: Optional[Dict[str, Any]] = None
    predicted_yield: float
    unit: str = "unconfirmed"
    risk: Dict[str, Any] = Field(default_factory=dict)
    uncertainty: Dict[str, Any] = Field(default_factory=dict)


class ScenarioResult(BaseModel):
    name: str
    input: Optional[Dict[str, Any]] = None
    changes: Dict[str, float]
    predicted_yield: float
    unit: str = "unconfirmed"
    risk: Dict[str, Any] = Field(default_factory=dict)
    uncertainty: Dict[str, Any] = Field(default_factory=dict)


class ScenarioComparison(BaseModel):
    absolute_change: float
    percentage_change: Optional[float] = None
    direction: str
    is_material: bool
    materiality_label: str
    material_threshold: float
    feature_diffs: Dict[str, Any]


class ScenarioResponse(BaseModel):
    baseline: ScenarioBaseline
    scenario: ScenarioResult
    comparison: ScenarioComparison
    progression: Dict[str, Any]
    explanation: Dict[str, Any]
    validation: Dict[str, Any]
    interpretation: Dict[str, Any]


class SensitivityRequest(BaseModel):
    current_input: FarmInput = Field(...)
    feature: str = Field(..., example="soil_moisture")
    num_points: int = Field(default=10, ge=3, le=50)


class SensitivityPoint(BaseModel):
    feature_value: float
    predicted_yield: float
    difference: float


class SensitivityResponse(BaseModel):
    feature: str
    display_name: str
    unit: str
    baseline_value: Optional[float] = None
    baseline_yield: float
    points: List[SensitivityPoint]
    disclaimer: str


class PresetsRequest(BaseModel):
    current_input: FarmInput = Field(...)
    preset_type: Optional[str] = Field(
        default=None,
        description="Optional specific preset name (e.g., improve_moisture, drought_stress)",
    )


class PresetsResponse(BaseModel):
    presets: Dict[str, Any]

File: frontend/src/types/scenario.ts
import { FarmInput } from './farm';
import { RiskPayload, UncertaintyPayload } from './prediction';
import { FeatureContribution } from './explanation';

export interface ScenarioRequest {
  current_input: FarmInput;
  changes: Record<string, number>;
  scenario_name?: string;
  run_explanation?: boolean;
}

export interface ScenarioBaseline {
  input?: Record<string, unknown>;
  predicted_yield: number;
  unit: string;
  risk: RiskPayload;
  uncertainty: UncertaintyPayload;
}

export interface ScenarioResult {
  name: string;
  input?: Record<string, unknown>;
  changes: Record<string, number>;
  predicted_yield: number;
  unit: string;
  risk: RiskPayload;
  uncertainty: UncertaintyPayload;
}

export interface FeatureDiffItem {
  baseline: number;
  scenario: number;
  difference: number;
}

export interface ScenarioComparison {
  absolute_change: number;
  percentage_change?: number | null;
  direction: 'increase' | 'decrease' | 'no_material_change';
  is_material: boolean;
  materiality_label: string;
  material_threshold: number;
  feature_diffs: Record<string, FeatureDiffItem>;
}

export interface ScenarioResponse {
  baseline: ScenarioBaseline;
  scenario: ScenarioResult;
  comparison: ScenarioComparison;
  progression: {
    risk: {
      baseline_level: string;
      scenario_level: string;
      risk_changed: boolean;
    };
    uncertainty: {
      baseline_level: string;
      scenario_level: string;
      uncertainty_changed: boolean;
    };
  };
  explanation: {
    changed_features: string[];
    delta_contributors: FeatureContribution[];
  };
  validation: {
    within_training_range: boolean;
    distance?: number;
    distance_class?: string;
    warnings: string[];
  };
  interpretation: {
    summary: string;
    scenario_reliability: 'HIGH' | 'MEDIUM' | 'LOW';
    reliability_reason?: string;
    causal_claim: boolean;
  };
}

export interface SensitivityPoint {
  feature_value: number;
  predicted_yield: number;
  difference: number;
}

export interface SensitivityResponse {
  feature: string;
  display_name: string;
  unit: string;
  baseline_value?: number;
  baseline_yield: number;
  points: SensitivityPoint[];
  disclaimer: string;
}

export interface SensitivityRequest {
  current_input: FarmInput;
  feature: string;
  num_points?: number;
}

export interface PresetsResponse {
  presets: Record<string, Record<string, number>>;
}
